akka-docs/rst/scala/code/docs/akka/typed/IntroSpec.scala

```diff
@@ -0,0 +1,206 @@
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.concurrent.Await
+//#imports
+import akka.testkit.AkkaSpec
+import akka.typed.TypedSpec
+
+object IntroSpec {
+
+  //#hello-world-actor
+  object HelloWorld {
+    final case class Greet(whom: String, replyTo: ActorRef[Greeted])
+    final case class Greeted(whom: String)
+
+    val greeter = Static[Greet] { msg ⇒
+      println(s"Hello ${msg.whom}!")
+      msg.replyTo ! Greeted(msg.whom)
+    }
+  }
+  //#hello-world-actor
+
+  //#counter-actor
+  object Counter {
+
+    sealed trait Command
+    final case class GetAndAdd(amount: Int, replyTo: ActorRef[CurrentCount]) extends Command
+    final case class AddAndGet(amount: Int, replyTo: ActorRef[CurrentCount]) extends Command
+
+    final case class CurrentCount(x: Int)
+
+    val counter: Behavior[Command] = behavior(0)
+
+    private def behavior(count: Int): Behavior[Command] =
+      Total {
+        case AddAndGet(amount, replyTo) ⇒
+          val nextCount = count + amount
+          replyTo ! CurrentCount(nextCount)
+          behavior(nextCount)
+        case GetAndAdd(amount, replyTo) ⇒
+          replyTo ! CurrentCount(count)
+          behavior(count + amount)
+      }
+  }
+  //#counter-actor
+
+  //#chatroom-actor
+  object ChatRoom {
+    //#chatroom-protocol
+    sealed trait Command
+    final case class GetSession(screenName: String, replyTo: ActorRef[SessionEvent])
+      extends Command
+    //#chatroom-protocol
+    //#chatroom-behavior
+    private final case class PostSessionMessage(screenName: String, message: String)
+      extends Command
+    //#chatroom-behavior
+    //#chatroom-protocol
+
+    sealed trait SessionEvent
+    final case class SessionGranted(handle: ActorRef[PostMessage]) extends SessionEvent
+    final case class SessionDenied(reason: String) extends SessionEvent
+    final case class MessagePosted(screenName: String, message: String) extends SessionEvent
+
+    final case class PostMessage(message: String)
+    //#chatroom-protocol
+    //#chatroom-behavior
+
+    val behavior: Behavior[GetSession] =
+      ContextAware[Command] { ctx ⇒
+        var sessions = List.empty[ActorRef[SessionEvent]]
+
+        Static {
+          case GetSession(screenName, client) ⇒
+            sessions ::= client
+            val wrapper = ctx.spawnAdapter {
+              p: PostMessage ⇒ PostSessionMessage(screenName, p.message)
+            }
+            client ! SessionGranted(wrapper)
+          case PostSessionMessage(screenName, message) ⇒
+            val mp = MessagePosted(screenName, message)
+            sessions foreach (_ ! mp)
+        }
+      }.narrow // only expose GetSession to the outside
+    //#chatroom-behavior
+  }
+  //#chatroom-actor
+
+}
+
+class IntroSpec extends TypedSpec {
+  import IntroSpec._
+
+  def `must say hello`(): Unit = {
+    //#hello-world
+    import HelloWorld._
+    import scala.concurrent.ExecutionContext.Implicits.global
```

Comment:
What is the alternative? We cannot use the ActorSystem’s dispatcher since we want to run callbacks after that is shut down.